### PR TITLE
EventManager refactor for version 3

### PR DESCRIPTION
**Pull request by weierophinney**
This patch refactors zend-eventmanager with the following goals:
- **Maintainability.**
  - It removes methods where they have no clear purpose or are redundant.
  - It simplifies signatures where arguments have multiple possible types or meanings; if necessary, it creates new methods when multiple signatures are deemed useful.
- **Performance.**
  - Queues are inlined into the implementations whenever possible to promote performance.
  - Arrays are preferred over objects for queues due to the speed of array manipulation in PHP.
  - Features such as lazy listeners were added to help users further optimize their event manager usage.
  - A benchmarking suite using Athletic was written to help gauge the differences between versions and approaches.
- **Usability.**
  - Whenever possible, signatures were refactored to have a single set of valid arguments, and a single execution path, providing reliability and easing the ability to learn usage.
  - Documentation was written alongside the implementation to aid in identifying difficult and redundant usage areas.

Alongside these goals, we also tried to minimize migration for end users. To this end, [migration document was written](https://github.com/weierophinney/zend-eventmanager/tree/8b6844b895c4ee1747aa9f86b7c7d40284eb5cd3/doc/book/migration), and a 2.6.0 release is planned in order to deprecate methods as well as add implementation methods to the `EventManager` that will aid in helping users prepare their code for version 3.

Listeners are executed exactly as they were in version 2, using the same priorities and precedence between local explicit and wildcard listeners, and shared and shared wildcard listeners. Tests were added to describe and document these rules.
### Removed functionality
- `GlobalEventManager` and `StaticEventManager` are removed (with prejudice!).
- `ProvidesEvents`, which was previously deprecated, is removed.
- `EventManagerInterface::setSharedManager()` is removed. Shared managers are now expected to be injected during instantiation.
- `EventManagerInterface::getEvents()` and `getListeners()` are removed; they had now purpose within the implementation.
- `EventManagerInterface::setEventClass()` was renamed to `setEventPrototype()`, which now expects an `EventInterface` instance. That instance will be cloned whenever a new event is created.
- `EventManagerInterface::attachAggregate()` and `detachAggregate()` are removed. Users should use the `attach()` and `detach()` methods of the aggregates themselves.
- `SharedEventAggregateAwareInterface` and `SharedListenerAggregateInterface` are removed. This was an undocumented and largely unused feature.
- `SharedEventManagerAwareInterface` is removed. A new interface, `SharedEventsCapableInterface` defines the `getSharedManager()` method from the interface, and `EventManagerInterface` extends that new interface.
- `SharedEventManagerInterface::getEvents()` is removed, as it had no purpose in the implementation.
### Changed signatures
- `EventManager::__construct()` now accepts an optional `SharedEventManagerInterface` instance as the first argument, and an optional array of identifiers as the second. As identifiers have no meaning without a shared manager present, they are secondary to providing the shared manager.
- `EventManagerInterface::trigger()` changes its signature to `trigger($eventName, $target = null, $argv = [])`; each argument has exactly one possible meaning; the `$eventName` can only be a string event name. The fourth `$callback` argument is removed.
- `EventManagerInterface::triggerUntil()` changes its signature to `triggerUntil(callable $callback, $eventName, $target = null, $argv = null)`. Each argument has exactly one meaning.
- `EventManagerInterface` adds two new methods for triggering provided `EventInterface` arguments: `triggerEvent(EventInterface $event)` and `triggerEventUntil(callable $callback, EventInterface $event)`.
- `EventManagerInterface::attach()` and `detach()` change their signatures to `attach($eventName, callable $listener, $priority = 1)` and `detach(callable $listener, $eventName = null)`, respectively. Note that `$eventName` can now only be a string event name, not an array or `Traversable`.
- `EventManagerInterface::setIdentifiers()` and `addIdentifiers()` change their signatures to each only accept an _array_ of identifiers.
- `SharedEventManagerInterface::getListeners()` changes signature to `getListeners(array $identifiers, $eventName)` and now guarantees return of an array. Note that the second argument is now _required_.
- `SharedEventManagerInterface::attach()` changes signature to `attach($identifier, $eventName, callable $listener, $priority = 1)`. The `$identifier` and `$eventName` **must** be strings.
- `SharedEventManagerInterface::detach()` changes signature to `detach(callable $listener, $identifier = null, $eventName = null)`; `$identifier` and `$eventName` **must** be strings if passed.
- `ListenerAggregateInterface::attach()` adds an optional `$priority = 1` argument. This was used already in v2, but not dictated by the interface.
- `FilterInterface::attach()` and `detach()` have changed signature to `attach(callable $callback)` and `detach(callable $ilter)`, respectively.
- `FilterIterator::insert()` has been modified to raise an exception if the value provided is not a callable.
- `ResponseCollection::setStopped()` no longer implements a fluent interface.
### New functionality
- `LazyListener` allows wrapping:
  - fetching a listener service from a container-interop container, and
  - invoking a designated listener method with the provided event.
- `LazyEventListener` extends `LazyListener`, and provides metadata for discovering the intended event name and priority at which to attach the lazy listener; these are consumed by:
- `LazyListenerAggregate`, which, provided a list of `LazyEventListeners` and/or definitions to use to create them, acts as an aggregate for attaching a number of such listeners at once.
## Benchmarks

Benchmarks are listed in a [gist](https://gist.github.com/weierophinney/34d5781d1a9f34785c64). In particular:
- [benchmark of version 2](https://gist.github.com/weierophinney/34d5781d1a9f34785c64#file-benchmarks-v2-log)
- [benchmark of final version of this implementation](https://gist.github.com/weierophinney/34d5781d1a9f34785c64#file-benchmarks-v3-internal-queue-log)

The highlight is the MultipleEventMultipleLocalAndSharedListener benchmark, which demonstrates a real-world usage scenario in zend-mvc with many local and shared listeners, and many events being triggered. This benchmark jumped from 435 operations per second in version 2 to 1265 operations per second in this implementation — a change of ~300%! Considering this represents 10–50X the number of events triggered in a reasonably sized zend-mvc application, performance will be in the sub-millisecond range.

### CODE REVIEW

**Comment by weierophinney** [comment]
Ping @akrabat @Ocramius @bakura10 @codeliner - this is an alternate approach that retains BC, but improves performance radically.

**Comment by bakura10** [comment]
I will have a look at the code but does this PR contains the lazy listener feature of my refactor? This is actually what I consider the most important feature of my refactor and is the key for intensive use of Evm without having to revert to complex solutions.

**Comment by weierophinney** [comment]
Not yet, but I'd like to get that feature in; I think it would simplify
listener attachment a ton.

**Comment by gianarb** [comment]
_178 additions and 386 deletions._
This PR starts well :+1:

**Comment by bakura10** [comment]
A lot more could be removed imho:
- $eventClass (what's the point of it? assuming Zend\EventManager\Event by default is enough, and remove one confusing method and property... and if they need to override it, they can do it using a custom event, which is better).
- the trigger method is still awfully complex. What about this:

``` php
 public function trigger($event, $target = null, $argv = array(), callable $callback = null)
    {
        if ($event instanceof EventInterface) {
            $e        = $event;
            $event    = $e->getName();
            $callback = $target;
        } else {
            $e = new Event();
            $e->setName($event);
            $e->setTarget($target);
            $e->setParams($argv);
        }

        // Initial value of stop propagation flag should be false
        $e->stopPropagation(false);
        return $this->triggerListeners($e, $callback);
    }
```
- prepareArgs: is this really needed? I think we should encourage people to create their event object instead of relying on this.

In overall I'm not against keeping the SEM, but for me the problem is that currently, the use case for SEM  is what most people want but it is complicated to use. If SEM stays here for most complex usage, why not, but the most simple use case should be the default one. That's why I think EVM SHOULD be a shared service, to simply allow attaching without having to revert to the complexity of SEM.

**Comment by codeliner** [comment]
@weierophinney :+1: for the performance improvments. That + BC is a strong argument for keeping the SEM.

But I see problems with removing `EVM::addIdentifiers` because

``` php
class MyClass extends AbstractEventTriggeringClass
{
    public function setEventManager(EventManagerInterface $evm) {
        parent::setEventManager($evm);
        $evm->addIdentifiers([MyClass::class]);
    }
}
```

... is no longer possible and adds even more complexity to the game because now `MyClass` needs to be aware of the identifiers used in `AbstractEventTriggeringClass` and all the way down.

# AND

> Any shared listeners attached after the instance is created will not be executed. This means attachment of shared listeners should be done early, preferably during application bootstrap, but really anytime before an instance is fetched from the service manager.

That sounds like a debugging nightmare! How do you want to document it so that EVERY developer is aware of it?
I'm not completely against it but the question is for whom the zend framework is made for?  A small group of experts who always know what they are doing or a larger group of PHP developers who are fighting against deadlines, inaccurate requirements, ... every day and therefor need tools / a framework / components which just work!

**Comment by weierophinney** [comment]
@codeliner Thanks for the feedback! I've updated to re-add `addIdentifiers()`, which necessitated deferring the attachment of shared listeners to the first `trigger()` operation. This actually solves a lingering doubt I had with the original refactor, as aggregating shared listeners early leads to some potential order-of-operations issues (which you confirmed in your comment). This also simplified the logic for aggregating wildcard events — and posed no changes to the initial performance improvements I observed!

(However, if you attempt to add identifiers or a shared manager after the first `trigger` operation an exception is now raised. I consider those situations to be edge cases in the current implementation, and feel raising the exception is a reasonable change at this time.)

It also addresses your second point. It's rare (in fact, I cannot think of any time it happens in the current MVC) that either identifiers or additional shared listeners will be added _after_ you have triggered the first event. However, it's not uncommon that you'll still be setting up both after you initialize the EM instance. As such, the changes in the latest commit address those issues.

**Comment by weierophinney** [comment]
> $eventClass (what's the point of it? assuming Zend\EventManager\Event by default is enough, and remove one confusing method and property... and if they need to override it, they can do it using a custom event, which is better).

The point of it is to allow a default event implementation _per instance_; as an example, a DB component could have a DbEvent that has additional methods and/or defaults. However, that could also be accomplished by sub-classing the EM.

> the trigger method is still awfully complex.

Your recommendation looks reasonable, but before I incorporate it, I'm going to survey how `trigger()` is called in a number of projects to make sure this wouldn't be removing use cases that are more widely used.

> prepareArgs: is this really needed? I think we should encourage people to create their event object instead of relying on this.

The idea behind `prepareArgs()` is to create an _object_ containing the event arguments, which allows updating the individual parameters by reference. It's actually used in a fair number of places, but it may be something that can be accomplished better in other ways.

**Comment by akrabat** [comment]
I like the approach. Will try and play with the code on Sunday / early next week.

**Comment by codeliner** [comment]
@weierophinney :+1: 

> However, if you attempt to add identifiers or a shared manager after the first trigger operation an exception is now raised. I consider those situations to be edge cases in the current implementation, and feel raising the exception is a reasonable change at this time.

Agreed. At least as long as you use PHP's shared nothing "feature". Running an app in a node.js like event loop (f.e. with reactphp) is a different story. In this case I can imagine that you may want to `detach` a listener (which is not possible with this refacor!) and add another one -> hot plug in.
But that's future. Don't know if you plan to support event loops with zend-diactoros. After moving some ideas from node to PHP this would be the next logical step for me.

File: src/EventManager.php [hunk 1]
```diff
@@ -174,8 +164,10 @@ public function addIdentifiers($identifiers)
      * @return ResponseCollection All listener return values
      * @throws Exception\InvalidCallbackException
      */
-    public function trigger($event, $target = null, $argv = [], $callback = null)
+    public function trigger($event, $target = null, $argv = array(), callable $callback = null)
```
Comment (samsonasik) [inline src/EventManager.php]: <code>$argv = array()</code> can be <code>$argv = []</code>

Comment (weierophinney) [inline src/EventManager.php]: I'll run it through the new CS rules when complete; right now, I'm mainly
testing some concepts for refactoring.

**Comment by weierophinney** [comment]
> In this case I can imagine that you may want to detach a listener (which is not possible with this refacor!) and add another one -> hot plug in.
> But that's future.

This would not be hard to accomplish at this point. The logic would be far easier than we have currently, as you only need the callable.

> Don't know if you plan to support event loops with zend-diactoros. After moving some ideas from node to PHP this would be the next logical step for me.

Diactoros doesn't need them, really; its primary focus is implementation of the HTTP message interfaces. (The `Server` implementation in Diactoros is essentially a nod at the front-controller pattern currently.) I'd put any idea of an event loop in a different package, to be honest. But it is definitely something we want to target eventually.

**Comment by weierophinney** [comment]
Odd failure on HHVM:
- https://travis-ci.org/zendframework/zend-eventmanager/jobs/80916792#L214

However, tests passed (segfault was within PHPUnit _after_ test completion).

At this point, we have excellent test coverage. Creating a few TODO items, and then we can start reviewing for merge.

**Comment by weierophinney** [comment]
I took a look at the "lazy listener" functionality as proposed by @bakura10, and I'm not convinced by the implementation. It essentially requires:
- A static class method, `attachAggregate()`
- In which it will attach using pseudo-callback notation: `[self::class, 'onWhatever']`
- and the EventManager will determine it's non-callable at `trigger()` and instantiate the listener class.

I have problems with this for a few reasons:
- I'd rather we hint on `callable` for the listeners. This makes the code more robust, as we can push validation of the argument to PHP itself, instead of having to validate it.
- It leads to complex logic when iterating listeners; we have to determine if it's callable, and, if not, try and create something that is out of the provided listener. Then, for performance, we also need to cache it.
- Finally, and the biggest nail in the coffin as far as I'm concerned, we lose a primary benefit of aggregate listeners: they allowed us to have objects _with dependencies_ that we could attach as listeners. For example, in Apigility, we have a number of listeners with dependencies on things like the Authentication component, a database, etc.; these are defined as services, and attached as aggregates to the EventManager, giving us stateful listeners. While this can be accomplished without aggregates, it seems ridiculous to change aggregates to only allow static listeners.

**However**, I think we _already_ can achieve lazy listeners, due to something I've already alluded to: listeners _must_ be callables. This allows creating lazy listeners via closures!

``` php
$events->attach('foo', function ($e) use ($container) {
    return $container->get('MyCustomListener')->onWhatever($e);
});
```

Essentially, if we want lazy listeners, we can wrap the logic in closures, pushing the performance hit to when the listener is actually triggered. With PHP 7, it would be relatively simple to achieve lazy aggregates as well via anonymous classes.

As such, I'm ticking off the box "lazy services", and will move to the other items on the list.

**Comment by weierophinney** [comment]
@codeliner — I've looked into what it would take for the `detach()` logic, and it's _very_ complex, as it turns out. It would require us to store the priority for each listener internally alongside the listener so that we can rebuild the priority queue after locating and removing the given listener.

I _have_ implemented `EventManager::clearListeners($event)` and `SharedEventManager::clearListeners($identifier, $event = null)`, however, and my inclination is that if you _do_ need to worry about removing listeners on-demand, you could do so by fetching the listeners (using `getListeners($event)`), clearing listeners, removing members from the list, and re-injecting in the order in which they were returnd to you. It's ugly, but works:

``` php
// We'll assume `$callback` is a listener we want to remove:
$listeners = $events->getListeners('event-of-interest');
$events->clearListeners('event-of-interest');
foreach ($listeners as $listener) {
    if ($listener === $callback) {
        continue;
    }
    $events->attach('event-of-interest', $listener);
}
```

It will work:
- During first `trigger()`, we attach shared listeners directly to the EM instance; this means that `getListeners()` will return the _shared_ listeners as well after that first `trigger()`.
- After first `trigger()`, we never re-attach shared listeners.
- Clearing listeners will thus remove both local and shared listeners on that event.
- However, you can always add local listeners.

So, the above will work. Where it gets dicey is priority.

You'll be iterating them in priority order, so re-attaching them preserves the original order. The problem is if you attach _other_ listeners later, using priority: at this time, they'll all be added with priority 1; we will have no record of the original priority. As such, any priority you specify later when attaching listeners will be relative to the default of 1.

And that's why I don't want to re-add this facility; it requires we store the priority alongside the listener, inflating storage, making getting at the listeners harder (we're essentially back to the `CallbackHandler` solution), and making the internals overall far more complex, leading us back to performance issues.

tl;dr: Essentially, if you want a solution that has both priority and the ability to detach, you lose performance.

**Comment by bakura10** [comment]
@weierophinney , actually the hit is when you register aggregate listener. Aggregate listener is the one that do the attach, but the dependencies are part of the aggregate listener. This is where you have the performance it.

This would mean wrapping the aggregate listener, but if you do that, the aggregate listener do not have the chance to register its listeners.

**Comment by weierophinney** [comment]
@bakura10 Attaching has very little measurable performance impact, though (particularly with the revisions in this PR); where are you arriving at the idea that it incurs a performance hit?

**Comment by bakura10** [comment]
The expensive part is creating all the listener aggregates, which all needs to pull many dependencies from service manager. This is what incurs a performance hit, not the attaching itself :).

**Comment by bakura10** [comment]
When you do that:

``` php
$aggregate = $sm->get(UserListenerAggregate::class);
$aggregate->attach();
```

This is the "get" that is expensive. But becuase you need to create the lsitener aggregate to bind events, you have no other choice. That's what my PR tried to tackled with the static stuff :). Being able to attach events WITHOUT the expesnive part of creating the aggregate

**Comment by weierophinney** [comment]
@bakura10 —

> When you do that:
> 
> ``` php
> $aggregate = $sm->get(UserListenerAggregate::class);
> $aggregate->attach();
> ```
> 
> This is the "get" that is expensive. But becuase you need to create the lsitener aggregate to bind events, you have no other choice.

Understood. The static approach doesn't fix that, however; if you use that approach, _of course_ it will be faster, because you're not fetching anything from the service manager! And that makes it less usable, because you can no longer have aggregate listeners that have dependencies.

Which brings me back to my earlier suggestion: if the aggregate is only attaching to a single event, wrapping the logic of pulling it from the container will work _now_; just bypass the `attachAggregate()` call entirely, and attach the closure directly to the EM.

For _true_ aggregates that attach many listeners, you would need a different approach. But that can be done in userland as well, with a little finesse:

``` php
class ThisIsAnAggregate implements ListenerAggregateInterface
{
    public static $attachTo = [
        [ 'event' => 'event1', 'method' => 'onEvent1', 'priority' => 1000 ],
        [ 'event' => 'event2', 'method' => 'onEvent2', 'priority' => -100 ],
        [ 'event' => 'event3', 'method' => 'onEvent3', 'priority' => 1 ],
    ];

    // define a constructor with dependencies ...

    public function attach(EventManagerInterface $events, $priority = 1)
    {
        foreach (self::$attachTo as $spec) {
            $events->attach($spec['event'], [ $this, $spec['method'] ], $spec['priority']);
        }
    }

    // define the event listener methods ...
}

// In your application:
foreach (ThisIsAnAggregate::$attachTo as $spec) {
    $events->attach($spec['event'], function ($e) use ($services) {
        return $services->get(ThisIsAnAggregate::class)->{$spec['method']}($e);
    }, $spec['priority']);
}
```

It makes use of statics, just like what you were doing, but doesn't require any changes to any existing interfaces, either in v2 or this proposal. By using a closure, we can put anything we want into the logic, including any environmental aspects we may want to use to decide whether or not we _really_ need to pull the service.

I suppose we could do a new `attachLazy()` method, which would accept a class name and optionally a list of context variables to bind to closures.
- The class would need to implement a new interface, which
- Defines a static `getAttachSpecifications()` method. It would accept the context variables to which to bind closures, and would return an array of:
- `EventListenerSpecification` instances. These would have methods for retrieving the event, callback, and priority.

and then `attachLazy()` would attach each of the returned specifications.

Or we could just document the solution I indicated above somewhere. :smile: Thoughts, @bakura10 ?

**Comment by bakura10** [comment]
Hhhmmm yeah that sounds like a nice workaround. It would be nice if we could have some syntactic sugar around this, but it should work.

**Comment by bakura10** [comment]
How the FastPriorityQueue compared to my simpler, dependency-less implementation here: https://github.com/zendframework/zend-eventmanager/pull/2?

I've been able to do that with pure array and much simpler code. My code being much much smaller, maybe you could start from this base, and start re-adding features you need (actually, only the sharedEventManager is really missing from my impl, and a few methods you may want for compatibility).

**Comment by Maks3w** [comment]
I would like to see removed all those setters on the EventInterface.

I think trigger() can typehint EventInterface without problems.

``` php
$eventManager->trigger($name, $target, $params, $callback);
$eventManager->trigger(new Event($name, $target, $params), $callback);
```
1. Still being simple trigger events without specialized classes.
2. Remove all that magic stuff about modify the object.
3. Remove those setters in the interface.

The only thing its no possible reuse an event with different event names (Is this really a need?)

``` php
$eventManager->attach('event.foo', function(EventInterface $eventFoo) {
   $eventManager->trigger('event.BAZ', $eventFoo);
});
```

**Comment by weierophinney** [comment]
Look at it yourself on the stdlib development branch. It's standalone, and
offers ~5x faster performance than SplPriorityQueue. @ezimuel identified
the queue as a bottleneck, and created this as a solution. Internally, it
uses priority as keys, with plain array values.

**Comment by weierophinney** [comment]
On Sep 18, 2015 5:38 AM, "Maks3w" notifications@github.com wrote:

> I would like to see removed all those setters on the EventInterface.

That would break BC. The ability to set parameters, in particular, needs to
be kept.

> I think trigger can typehint EventInterface without problems.
> 
> $eventManager->trigger($name, $target, $params, $callback);
> $eventManager->trigger(new Event($name, $target, $params), $callback);
> 
> Still being simple trigger events without specialized classes.
> Remove all that magic stuff about modify the object.
> Remove those setters in the interface.
> 
> The only thing its no possible reuse an event with different event names
> (Is this really a need?)

I've used it a number of times, and we use it in Apigility for resource
events.

That said, it could be done other ways. But if we can keep existing
functionally, while improving performance, if prefer that.

**Comment by Maks3w** [comment]
Ok, I really think its better don't break so many changes at a time but I think then we can mark the signature as deprecated.

May we could even define a new method `triggerEvent(EventInterface, callable)`

**Comment by Maks3w** [comment]
PD: Remove the setters from the interface is not a BC Break

**Comment by weierophinney** [comment]
On Sep 18, 2015 6:28 AM, "Maks3w" notifications@github.com wrote:

> PD: Remove the setters from the interface is not a BC Break

It absolutely is, because calling the methods from existing code bases
using new implementations will now raise errors.

**Comment by Maks3w** [comment]
How many implementations use the setters outside of the EventManager class?. I think the potentional break is very reduced compared with the rest of breaks proposed here.

**Comment by Maks3w** [comment]
Could you split the PR in different topics?
- Toolkit
- callable typhint
- StaticEventManager removal
- Removal of deprecations
- ...

I see its possible start to merge many things

**Comment by weierophinney** [comment]
On Sep 18, 2015 7:44 AM, "Maks3w" notifications@github.com wrote:

> How many implementations use the setters outside of the EventManager
> class?.

zend-mvc and Apigility for starters, which should be enough to justify
keeping them.

**Comment by weierophinney** [comment]
@Maks3w —

> Could you split the PR in different topics?

No. The amount of effort required to do that is ridiculously large, and would actually impede the refactoring effort, and, to my mind, review as well.

The goals of this PR are, in order:
- Improve performance as much as possible.
- Keep as much BC as possible with regards to signatures; break BC where it makes sense for performance, usability, and/or maintenance.
- Introduce features only where they aid in the above two goals.

Going over the changes:
- Removal of StaticEventManager aids in usability (one less signature to learn) and maintenance (less code, removal of global state).
- Callable typehint aids in performance and usability. This leads to the one real BC break: `attach()` no longer accepts aggregates. That, however, leads to better separation of concerns, aiding performance and maintenance.
- Removal of deprecations and deprecated behavior aids in maintenance, usability, and performance. _Most of these removals happened as a result of refactoring for performance_, and, as such, are incidental more often than intentional. Discussing them separately would make no sense.
- Toolkit changes help us verify changes more easily (aiding maitenance).
- Re-adding `detach()` will both remove and create a BC break (as the signature will need to change due to no longer using `CallbackHandler`). This will affect usability (you'll be able to pass the event name, which is a new feature), performance (the changes to use FastPriorityQueue offer > 400X performance gains for `remove()` operations!), and maintenance (we'll have an easier, more maintainable mechanism for this operation).
- Adding lazy listeners will aid performance and usability.

Everything else is essentially "under the hood," and has no effect on end users. Many of the above concerns overlap, and breaking them into separate PRs would make review far more complicated.

**Comment by weierophinney** [comment]
> The only thing its no possible reuse an event with different event names (Is this really a need?)

This is used extensively in both the `ModuleManager` and MVC implementations; the same `ModuleEvent` and `MvcEvent` instances are pushed through the entire workflow of those two components, and the new name is injected for each new event triggered. So, definitely a need; we use it _within the framework_!

**Comment by weierophinney** [comment]
@bakura10 — This is for you:
- https://github.com/zendframework/zend-eventmanager/commit/1eea094084a8c27d7fa4034490c462dfcd03573a

I found a solution for the lazy listeners. :smile:

**Comment by weierophinney** [comment]
Updated the issue description with the current set of changes, as well as an updated TODO list; visit the pull request if you would like to see what remains.

**Comment by danizord** [comment]
@weierophinney I think the `LazyListener` could be the callable itself. So that we won't need to call `LazyListener#getListener()` to build the callable.

**Comment by weierophinney** [comment]
I had the same thought last night, @danizord. :smile: Will push updates in
a few hours.

**Comment by weierophinney** [comment]
@danizord and @bakura10 — Incorporated your feedback. The lazy listener story is much simpler now:

``` php
$events->attach('foo', new LazyListener($definition, $container));
```

This also simplified the aggregate substantially.

I'm going to make one last change to this. Essentially, if you're creating a LazyListener to attach directly, you don't require the event argument, but currently it's required; similarly the priority argument is meaningless, even if provided. My plan is to have two separate classes, LazyListener and LazyEventListener. LazyListener will not require the "event" argument, while, LazyEventListener _will_. LazyListenerAggregate will require and/or build LazyEventListener instances.

**Comment by weierophinney** [comment]
@danizord and @bakura10 — d54022f implements the change to `LazyListener` and addition of `LazyEventListener`. I think that feature can now be considered complete.

**Comment by weierophinney** [comment]
I've got the benchmarks running now. Interestingly, the current version works against v2 without issue, which gives me some interesting performance data.

The worst case scenario (which is also essentially what zend-mvc illustrates) shows that the refactor is 37% faster; in fact, most scenarios where shared listeners were involved were in the range of 35-40% faster under the refactor.

Also of interest: the worst-case scenario triggers 5000 times, across a variety of shared and local listeners. With 5000 triggers, the total time spent by the EM was 3.4ms under v2, and 2.2ms under the refactor. My take on this is that _the EM itself is not and has not been the bottleneck, but that the refactor has significantly improved the situation regardless._

I've also run benchmarks with a patch that re-instates the "prepare listeners on every trigger" paradigm, in order to allow attaching shared listeners after the first trigger; considering this is the paradigm the current MVC operates under, it's an important consideration. In that scenario, the speeds are 10–30% slower than the current revision (e1e1441), but only under scenarios with shared listeners; thery're still faster than the v2 version by a factor of 30–35%. I'm going to try a few different approaches to see if I can reduce that to single digit differences.

**Comment by bakura10** [comment]
I won't accept a refactor that has less than 100% performance improvement :D.

More seriously, good work on that. I'll read that in more details this week end and I'll likely do a second pass on your refactor once it's merged to develop in order to see if I can squeeze more perf!

Envoyé de mon iPhone

**Comment by weierophinney** [comment]
@bakura10 — hahahahaha!

I've just pushed a change that allows preparation on each trigger. It does so in a smart way, however, and ends up being very close to the results prior to the change, and in some cases _faster_.

Which means we have feature parity, with much better performance!

**Comment by weierophinney** [comment]
Benchmarks are posted: https://gist.github.com/weierophinney/34d5781d1a9f34785c64

**Comment by weierophinney** [comment]
I've now updated the EventManagerInterface to define 4 discrete methods for triggering an event:
- `trigger($event, $target = null, $args = [])`
- `triggerUntil(callable $callback, $event, $target = null, $args = [])`
- `triggerEvent(EventInterface $event)`
- `triggerEventUntil(callable $callback, EventInterface $event)`

Each has exactly one execution path, simplifying both usage and maintenance, and, more interestingly, giving us yet another performance bump!

At this time, I'm going to turn to test coverage, and then documentation.

**Comment by bakura10** [comment]
I've just pulled your branch locally. I'll work on it soon but I think there are still lot of room for perf improvments :)

File: src/EventManager.php [hunk 2]
```diff
@@ -54,20 +64,25 @@ class EventManager implements EventManagerInterface
      *
      * @param  null|string|int|array|Traversable $identifiers
```
Comment (bakura10) [inline src/EventManager.php]: I'm not a fan of such permissive function. Can't we agree on only accepting array so that we can use array $identifiers = [] ?

**Comment by bakura10** [comment]
Should we remove fluent interface? Makes little sense in the EVM and is annoying to test, as usual :).

**Comment by weierophinney** [comment]
@bakura10 —

> I'd change the order to $event, $callback, $target, $argv, as the most important is the event.

and

> I'd change the order to $event, $callback

I considered those signatures, but decided against them. Here's my rationale: on `triggerUntil()`, that changes the relationship between the `$event`, `$target`, and `$argv` arguments, making order harder to remember. By putting the `$callback` at the front, we keep the original order. I also disagree with the assertion that the most important argument on the `*Until()` methods is the event; in those particular cases, the difference you're signaling in usage is that you have a condition that will cause execution to halt. That criteria is the diffentiating factor in the signature, and pushing it to the first argument makes that eminently clear.

**Comment by weierophinney** [comment]
> Should we keep the getEvents? 

I'm honestly uncertain. :) I don't have a use for it, except internally (I use it when attaching wildcard events). I'll definitely consider removing it.

**Comment by weierophinney** [comment]
> Should we remove fluent interface?

Yes. I'll make a note to do that.

File: src/EventManager.php (hunk 2 above)
Comment (weierophinney) [inline src/EventManager.php]: It's not supposed to; I've changed it already on my local copy to array only. (Since it delegates to `setIdentifiers()`, which also typehints on array at this time. You'll see that change shortly.

**Comment by weierophinney** [comment]
> I think there are still lot of room for perf improvments

I'd love to see them; I'm getting at most single digit improvements at this time with optimizations I'm working on. :wink:

**Comment by bakura10** [comment]
Same feedback for `getListeners`. I'm uncertain about having this in the interface. The interest of this is very limited. Why would you want to retrieve listeners for an event? To trigger? Then just use trigger ? :).

**Comment by weierophinney** [comment]
@bakura10 —

> I'm not sure about that. I often use an Event class that is used for several events. For instance:
> 
> `````` php
> class UserEvent extends Event
> {
>    const EVENT_CREATED = 'user.created';
>    const EVENT_UPDATED = 'user.updated';
> }
> And use:
> 
> ```php
> $evm->trigger(UserEvent::EVENT_CREATED, $userEvent);
> ``````
> 
> But your change no longer allow this because it relies on the fact that an event can only return a single name. And trigger does not support this neither because it expects $event to be a string.

What you're indicating is that you want `trigger()` to allow passing an event name followed by the event instance, which would implicitly set the event name on the instance.

I've specifically removed that capability. It provides too many paths through the method, making it harder to understand the possible signatures (there were something like 24 argument permutations before!), and thus harder to maintain.

Your code above would become:

``` php
$userEvent->setName(UserEvent::EVENT_CREATED);
$evm->triggerEvent($userEvent);
```

In other words, setting the event name becomes _explicit_, which is exactly what the change was intended to accomplish.

**Comment by bakura10** [comment]
Sounds good to me. Sold!

**Comment by weierophinney** [comment]
> Same feedback for getListeners. I'm uncertain about having this in the interface. The interest of this is very limited. Why would you want to retrieve listeners for an event? To trigger? Then just use trigger ? :).

I use it for testing at this point, but that can be accomplished in other ways. Will make a note to remove it.

**Comment by weierophinney** [comment]
@bakura10 — latest changes incorporate some of your feedback. Additionally, I had to change the logic in `fetchCurrentSharedListeners()` as testing showed it wasn't actually working as expected; the benefit, though, is another 15–20% performance boost!

**Comment by weierophinney** [comment]
I have restored test coverage, and improved upon it; additionally, the last few commits also improve performance even more, while keeping the key pieces of functionality that are unique to our offering, and which are required by the MVC.

Two items remain:
- Remove `getEvents()` and `getListeners()` from `EventManager(Interface)?` (per @bakura10 )
- Write migration documentation

I'll try and get these done in the next few days.

**Comment by bakura10** [comment]
Note for myself: we should rename method parameters to better match what they represent (ie. $eventName instead of $event). This also helps auto-completion.

**Comment by weierophinney** [comment]
I've written migration documentation, and documented the new features (lazy listeners). My plan is to migrate the rST documentation as well, and update it to follow the new API. At that point, we can consider it complete.

@bakura10 Do you want to do a pull request for the variable name changes you propose? You can issue it against my working branch, and I can merge it in to bring it into this patch.

**Comment by bakura10** [comment]
Naming changes are part of the PR I've made to your branch ;)

Envoyé de mon iPhone

**Comment by weierophinney** [comment]
I've merged some performance improvements from @bakura10 , and updated the benchmarks gist with the new benchmarks; they are now roughly 30X faster than v2!

In reviewing the changes he presented, I realized we could remove the `*Aggregate()` methods from `EventManagerInterface`, and should also add the `$priority` argument to `ListenerAggregateInterface::attach()`; I've added TODO items above for these two.

Additionally, now that these changes are in place, test coverage has once again decreased (despite adding many more tests to cover event prioritization), so I need to address that before we merge to develop.

**Comment by bakura10** [comment]
Did you try array_replace_recursive after all?

Anyway, thanks A LOT for taking care of the tests for me!

**Comment by bakura10** [comment]
Also, can you please remind me what's the FilterChain is about?

**Comment by weierophinney** [comment]
> Did you try array_replace_recursive after all?

I did. It was removing entries, unfortunately.

**Comment by weierophinney** [comment]
> Also, can you please remind me what's the FilterChain is about?

See https://github.com/weierophinney/zend-eventmanager/blob/feature/sem-refactor/doc/book/intercepting-filters.md

This was one of the use cases originally for the event manager. The implementation is a bit different than the standard event manager, but I haven't yet decided if it warrants its own component. Since it's lived in this component all along, easiest option is to keep it as optional functionality.

**Comment by weierophinney** [comment]
Test coverage is better now; only piece with sub-optimal coverage is the `EventManagerAwareTrait`.

**Comment by weierophinney** [comment]
Final changes are pushed. I've updated the description to reflect the final state of the refactor, including a synopsis of all changes, and an analysis of benchmarks. While the patch shows net double the lines, the majority of those are in tests and documentation.

I've also created #10 to track the deprecations we need to create for 2.6.0 to allow users time to prepare their code for 3.0. If anybody wants to tackle that, please, be my guest!

Please review! (pinging @bakura10 , @akrabat , @mwillbanks , and @ezimuel ).
